omegaconf/omegaconf.py

```diff
@@ -101,25 +101,55 @@ def SI(interpolation: str) -> Any:
 
 
 def register_default_resolvers() -> None:
-    def env(key: str, default: Any = _EMPTY_MARKER_) -> Any:
+    # # DEPRECATED: remove in 2.2
+    def legacy_env(key: str, default: Optional[str] = None) -> Any:
+        warnings.warn(
+            "The `env` resolver is deprecated, see https://github.com/omry/omegaconf/issues/573",
+        )
+
+        try:
+            return decode_primitive(os.environ[key])
+        except KeyError:
+            if default is not None:
+                # We handle the "null" string in a special way so as to preserve
+                # the old behavior of `env` while keeping the fix from #230.
+                if default.lower() == "null":
+                    return None
+                else:
+                    return decode_primitive(default)
+            else:
+                raise ValidationError(f"Environment variable '{key}' not found")
+
+    def env(key: str, default: Any = _EMPTY_MARKER_) -> Optional[str]:
         try:
-            val_str = os.environ[key]
+            return os.environ[key]
         except KeyError:
             if default is not _EMPTY_MARKER_:
+                if default is not None and not isinstance(default, str):
+                    raise ValidationError(
+                        f"The default value of the `oc.env` resolver must be a string or "
+                        f"None, but `{default}` is of type {type(default).__name__}"
+                    )
                 return default
             else:
                 raise ValidationError(f"Environment variable '{key}' not found")
 
-        # We obtained a string from the environment variable: we try to parse it
-        # using the grammar (as if it was a resolver argument), so that expressions
-        # like numbers, booleans, lists and dictionaries can be properly evaluated.
-        try:
-            parse_tree = parse(
-                val_str, parser_rule="singleElement", lexer_mode="VALUE_MODE"
+    def decode(expr: Optional[str]) -> Any:
+        """
+        Parse and evaluate `expr` according to the `singleElement` rule of the grammar.
+
+        If `expr` is `None`, then return `None`.
+        """
+        if expr is None:
+            return None
+
+        if not isinstance(expr, str):
+            raise ValidationError(
+                f"`oc.decode` can only take strings or None as input, "
+                f"but `{expr}` if of type {type(expr).__name__}"
             )
```

Comment:
This reminds me of Hydra [Functions class](https://github.com/facebookresearch/hydra/blob/44b843629e7f01e40b853bcdf66d7fa2645365fd/hydra/_internal/grammar/functions.py). It supports registration of functions and it uses the arguments types for automatic validation.
It's something we can think about as a future enhancement.

Hydra got a small set of function it supports, but at the same time we also support a limited set of input types here, so I think it's doable.

Reply:
Sounds like a good idea => created #612 to keep track of it